File: library/builtins/traits/module.rs
use crate::*;
use std::collections::HashMap;
use wipple::*;

#[derive(TypeInfo, Clone)]
pub struct Module {
    pub env: Env,
}

impl Primitive for Module {}

impl Module {
    pub fn new(env: Env) -> Self {
        Module { env }
    }
}

impl EvaluateBlockFn {
    pub fn into_module() -> Self {
        fn evaluate_block(block: &Block, env: &Env, stack: &Stack) -> Result<Value> {
            let env = env.child();
            env.set_evaluate_block(EvaluateBlockFn::into_module());

            block.reduce(&env, stack)?;
            Ok(Value::of(Module::new(env)))
        }

        EvaluateBlockFn::new(evaluate_block)
    }
}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    env.set_variable(stack, "Module", Value::of(Trait::of::<Module>()))?;

    // Module == Text
    env.add_text_relation::<Module>("module", stack)?;

    // Module == Function
    env.add_relation_between(stack, |module: Module| {
        Function::new(move |value, env, stack| {
            let name = value.get_or::<Name>("Expected a name", env, stack)?;
            name.resolve(&module.env, stack)
        })
    })?;

    // Module == Pattern
    env.add_relation_between_with(stack, |module: Module, stack| {
        let mut fields = HashMap::new();

        let variables = module.env.variables().0;
        for (name, variable) in variables {
            let pattern = variable
                .get_value(&module.env, stack)?
                .get_or::<Pattern>("Expected pattern", &module.env, stack)?
                .into_owned();

            fields.insert(name.clone(), pattern);
        }

        Ok(Pattern::new(move |value, env, stack| {
            let module = value
                .get_or::<Module>("Expected module", env, stack)?
                .into_owned();

            let validated_env = Env::new();

            let variables = module.env.variables().0;
            for (name, variable) in variables {
                let value = variable.get_value(env, stack)?;

                let validated = match fields.get(&name) {
                    Some(pattern) => pattern(&value, env, stack)?,
                    None => return Ok(None),
                };

                match validated {
                    Some(value) => validated_env.set_variable(stack, &name, value.into_owned())?,
                    None => return Ok(None),
                }
            }

            Ok(Some(Cow::Owned(Value::of(module))))
        }))
    })?;

    Ok(())
}

File: Library/Sources/Traits/Module.swift
import Wipple

public struct Module: Primitive {
    public var capturedEnv: Env

    public init(capturing env: Env) {
        self.capturedEnv = env
    }
}

public extension Module {
    static var evaluateBlock: EvaluateBlock {
        { block, value, env, stack in
            let env = env.child()
            env.evaluateBlock = Module.evaluateBlock

            try block.reduce(location: value.location, env, stack)

            return Value(Module(capturing: env))
        }
    }
}

internal func setupModule(_ env: Env, _ stack: Stack) throws {
    try env.setVariable("Module", to: Value(Trait(Module.self)), stack)

    // Module == Text
    try env.addRelation(text: "module", for: Module.self, stack)

    // Module == Function
    try env.addRelation(from: Module.self, to: Function.self, stack) { module in
        Function { value, env, stack in
            try value
                .get(Name.self, or: "Expected name", env, stack)
                .resolve(module.capturedEnv, stack)
        }
    }

    // Module == Pattern
    try env.addRelation(from: Module.self, to: Pattern.self, stack) { module, _, stack in
        let variables = module.capturedEnv.variables

        let fields = try variables.mapValues { variable in
            try variable
                .getValue(module.capturedEnv, stack)
                .get(Pattern.self, or: "Expected pattern", module.capturedEnv, stack)
        }

        return Pattern { value, env, stack in
            let module = try value.get(Module.self, or: "Expected module", env, stack)

            let matchedEnv = env.child()

            for (name, variable) in module.capturedEnv.variables {
                let value = try variable.getValue(env, stack)

                guard let pattern = fields[name],
                      let matchedValue = try pattern.match(value, env, stack)
                else {
                    return nil
                }

                try matchedEnv.setVariable(name, to: matchedValue, stack)
            }

            return Value(Module(capturing: matchedEnv))
        }
    }

    try env.setVariable(
        "New",
        to: Value(Function { value, env, stack in
            let trait = try value
                .evaluate(env, stack)
                .get(Trait.self, or: "Expected trait", env, stack)

            return Value(Pattern { value, env, stack in
                let module: Value = try Module.new(value, env, stack)
                return try .checked(trait: trait, value: module, env, stack)
            })
        }),
        stack
    )
}
